keps/sig-apps/85-Graduate-PDB-to-Stable/README.md

```diff
@@ -0,0 +1,319 @@
+The root of this issue, is that the rules in the disruption controller for what is a healthy pod, and the rules
+in the Eviction API for when a pod can be evicted without looking at the PDB are not the same. This means a pod can
+be considered unhealthy by the disruption controller so it does not count as healthy when computing `DisruptionsAllowed`,
+but will still require `DisruptionsAllowed` to be larger than 0 for it to be evicted. Some strange situations can
+arise from this. For example if we have a PDB with `MinAvailable  = 1` and 10 pods that are all in the CrashLoop state 
+(`Running`, but not `Ready`), we will not be allowed to evict any of the pods.
+
+This issue has been addressed with https://github.com/kubernetes/kubernetes/pull/94381. 
+
+#### Make the disruption controller more lenient for pods belonging to non-scale controllers
+
+The disruption controller is currently taking the very safe route whenever it encounters any
+issues with the targeted pods or their controllers. For all configurations of the PDB, except when
+`minAvailable` is a number, the PDB requires that it can find the controller and that
+the controller implements scale (either by being one of the built-in workloads or a CR where the CRD implements
+the scale subresource). If those conditions are not met for all pods targeted by the PDB, the disruption
+controller will set `DisruptionsAllowed` to 0, which means none of the pods can be evicted. There is an issue
+concerning this behavior: https://github.com/kubernetes/kubernetes/issues/77383.
+
+The current behavior of the disruption controller for the different types of input and the different
+types of pods that might be encountered are documented in: 
+https://docs.google.com/spreadsheets/d/12HUundBS-slA6axfQYZPRCeIu_Au_wsGD0Vu_oKAnM8/edit?usp=sharing
+
+##### Proposed solution
+
+To fix this, we will make two adjustments:
+
+Loosen the rules somewhat, so instead of just returning an error and block all evictions when
+the controller encounters invalid pods, it will ignore those pods when computing `AllowedDisruptions`. And
+similarly, the Eviction API will ignore the PDB when evicting those pods. This means that in a situation like
+was mentioned in the issue, where a PDB selector happen to match both pods belonging to a Deployment and pods
+belonging to a CronJob, the PDB will protect the Deployment pods as expected, but ignore the pods from the CronJob.
+
+Combined with loosening the rules in the controller, we also want to improve feedback to the user in these situations. 
+So while the pods from the CronJob described above will not be covered by the PDB,
+the controller will generate warning events for these situations that will provide accurate descriptions
+of what makes the pod ineligible for the PDB. In this case it would mean an event explaining that the pod
+has a controller that does not implement scale, and therefore can only be used if the PDB is using `minAvilable` as 
```

Comment:
When those events will be emitted? We don't want to suddenly significantly increase load on the control plane by sending gazillions of events.

Reply:
That is a good point. I have updated the language here and specified that we only should send the event the first time we encounter the situation rather than on every reconcile.